### PR TITLE
Use goblin as elf parser and only load segments on demand 

### PR DESCRIPTION
**Pull request by jessehui**
Libos now Only read in headers and needed segments. Also reduce a memory
copy.

### CODE REVIEW

File: src/libos/src/process/elf_file.rs [hunk 1]
```diff
@@ -1,95 +1,173 @@
...
+        }
+
+        let interpreter = if intepreter_count == 0 {
+            None
+        } else {
+            str::from_utf8(&elf_buf[intepreter_offset..intepreter_offset + intepreter_count]).ok()
+        };
+        println!("interpreter = {:?}", interpreter);
+        let elf_inner =
+            goblin::elf::Elf::lazy_parse_for_occlum(header, program_headers, elf_buf, interpreter)
```
Comment (guzongmin) [inline src/libos/src/process/elf_file.rs:84]: Is there another lazy parser? If it is the only one, we may remove the word occlum in the name.

File: src/libos/src/process/do_spawn/exec_loader.rs [hunk 2]
```diff
@@ -1,21 +1,22 @@
...
-    let file_buf = load_file_to_vec(&file_path, current_ref)?;
-    let is_script = is_script_file(&file_buf);
-    if is_script {
+) -> Result<(Option<String>, Arc<dyn INode>, Vec<u8>, ElfHeader)> {
+    let (inode, file_buf, elf_hdr) = load_file_hdr_to_vec(&file_path, current_ref)?;
+    if elf_hdr.is_some() {
+        Ok((None, inode, file_buf, elf_hdr.unwrap()))
+    } else {
+        // loaded file is not Elf format, try script file
+        if !is_script_file(&file_buf) {
```
Comment (guzongmin) [inline src/libos/src/process/do_spawn/exec_loader.rs:17]: Is the file_buf correctly initialized in this case?

File: src/libos/src/process/elf_file.rs (hunk 1 above)
Comment (jessehui) [reply src/libos/src/process/elf_file.rs]: Removed. Thank you for your review.

File: src/libos/src/process/do_spawn/exec_loader.rs (hunk 2 above)
Comment (jessehui) [reply src/libos/src/process/do_spawn/exec_loader.rs:22]: file_buf is always initialized in `read_elf64_as_vec` in deps/sefs/rcore-fs/src/vfs.rs.

**Comment by jessehui** [comment]
Changes for goblin have been merged (https://github.com/m4b/goblin/pull/254) and changes for SEFS can be found [here](https://github.com/occlum/sefs/pull/11).

**Review by guzongmin — APPROVED** [review]
LGTM

**Pull request merged**